### ISSUE TITLE
FR: information about episode length even before playback

### ISSUE THREAD

**Issue by woj-tek** [dyerc/Doughnut#45]
Currently we know what's the episode's length only once we start playing. It would be nice to have this info displayed for all episodes. From what I can see the information is enclosed in `<itunes:duration/>`

**Comment by woj-tek** [comment]
I decided to dig a little bit and as a result I created a PR #105. I tested it locally and it seems to be working fine.

(This is my first time I touched Swift code so not sure if it's perfectly correct way to handle it)

**Comment by GetToSet** [comment]
Assigning `iTunesDuration` (`<itunes:duration>`) is the correct approach but it could cause inconsistency in case the duration specified on the feed doesn't match the actual audio file.

Currently, we refresh the duration from the audio file when it finished loading (at `Player.onAssetLoadingFinished(avAsset: AVAsset, episode: Episode)`).

Imagine the following use case:
1. An incorrect episode duration read and saved to the `Episode` model after fetching the feeds.
2. A user begins to play that episode, the duration information will be corrected from the audio file.
3. A user then refreshes their feeds and the wrong episode duration will be saved again. Playback progress will become problematic until the user re-plays that episode.

I think we can review and merge this PR since it resolves a usability issue. I'll later fix this potential inconsistency before the next release. @dyerc what do you think?

**Comment by dyerc** [comment]
Thank you @woj-tek for submitting this PR! 

The inconsistencies @GetToSet mentioned are why I didn't use the iTunes duration originally. I wonder if we use the iTunes duration if the episode duration is currently nil (ie. a new episode detected or when this release is first deployed). Then the duration is updated in exactly the way it currently does. This way the true known duration would always take precedence, but the feed duration would be used initially and not ignored. What do you both think?

**Comment by woj-tek** [comment]
Or maybe simply have two fields/member: one with metadata duration and one with actual media duration? Probably somewhat less efficient memory wise, but probably the logic to handle it would be simpler?

**Comment by GetToSet** [comment]
> Or maybe simply have two fields/member: one with metadata duration and one with actual media duration? Probably somewhat less efficient memory wise, but probably the logic to handle it would be simpler?

This is actually what I thought. The `<itunes:duration>` is a property of the feed itself and a part of the specification, which should be better distinguished from actual playback duration (which is a state of the client application).

Here is my planned solution:

1. Use a separate property (probably `playbackDuration`) for the playback, which should also be persisted in the database.
2. For the episodes list UI, show `playbackDuration` when available, otherwise, show `duration`(`itunesDuration`).
3. On media asset loading finish, update `playbackDuration` if's not set. Otherwise, if `playbackDuration` don't match the current one, discard the `playPosition` since it signifies the underlying media resource has been changed.
4. On the episode editing modal panel, add a row for `duration`, and probably give the user a hint in case the duration doesn't match the `playbackDuration` and can probably be quickly corrected with a button click.

**Comment by woj-tek** [comment]
1-3 looks good IMHO. Not sure about 4 - I think it's sort of an edge case and should happen very rarely.

**Comment by GetToSet** [comment]
> 1-3 looks good IMHO. Not sure about 4 - I think it's sort of an edge case and should happen very rarely.

For the 4th point, I think Doughnut should allow professional users (e.g. podcasters) to edit every standard feed properties and even support exporting XML in the future. It's off the topic of this issue.

**Comment by GetToSet** [comment]
With some manual testing after #105, the current duration updating mechanism would be acceptable for most cases. This issue could be by now and I'll bring up a new one for discussion before working on further refinements.
